sdk/samples/iot/paho_iot_hub_pnp_component_example.c

```diff
@@ -0,0 +1,1004 @@
+{
+  int32_t required_size = az_span_size(mqtt_url_prefix) + az_span_size(iot_hub)
+      + az_span_size(mqtt_url_suffix) + (int32_t)sizeof(null_terminator);
+
+  if (required_size > destination_size)
+  {
+    return AZ_ERROR_INSUFFICIENT_SPAN_SIZE;
+  }
+
+  az_span destination_span = az_span_init((uint8_t*)destination, destination_size);
+  az_span remainder = az_span_copy(destination_span, mqtt_url_prefix);
+  remainder = az_span_copy(remainder, az_span_slice(iot_hub, 0, az_span_size(iot_hub)));
+  remainder = az_span_copy(remainder, mqtt_url_suffix);
+  az_span_copy_u8(remainder, null_terminator);
+
+  return AZ_OK;
+}
+
+// Read the user environment variables used to connect to IoT Hub
+static az_result read_configuration_and_init_client(void)
+{
+  az_span cert = AZ_SPAN_FROM_BUFFER(x509_cert_pem_file);
+  AZ_RETURN_IF_FAILED(
+      read_configuration_entry(ENV_DEVICE_X509_CERT_PEM_FILE, NULL, false, cert, &cert));
+
+  az_span trusted = AZ_SPAN_FROM_BUFFER(x509_trust_pem_file);
+  AZ_RETURN_IF_FAILED(
+      read_configuration_entry(ENV_DEVICE_X509_TRUST_PEM_FILE, "", false, trusted, &trusted));
+
+  az_span device_id_span = AZ_SPAN_FROM_BUFFER(device_id);
+  AZ_RETURN_IF_FAILED(
+      read_configuration_entry(ENV_DEVICE_ID, NULL, false, device_id_span, &device_id_span));
+
+  az_span iot_hub_hostname_span = AZ_SPAN_FROM_BUFFER(iot_hub_hostname);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      ENV_IOT_HUB_HOSTNAME, NULL, false, iot_hub_hostname_span, &iot_hub_hostname_span));
+
+  // Paho requires that the MQTT endpoint be of the form ssl://<HUB ENDPOINT>:8883
+  AZ_RETURN_IF_FAILED(
+      create_mqtt_endpoint(mqtt_endpoint, (int32_t)sizeof(mqtt_endpoint), iot_hub_hostname_span));
+
+  // Initialize the hub client with the hub host endpoint and the default connection options
+  az_iot_hub_client_options options = az_iot_hub_client_options_default();
+  options.model_id = model_id;
+  AZ_RETURN_IF_FAILED(az_iot_hub_client_init(
+      &client,
+      az_span_slice(iot_hub_hostname_span, 0, (int32_t)strlen(iot_hub_hostname)),
+      az_span_slice(device_id_span, 0, (int32_t)strlen(device_id)),
+      &options));
+
+  return AZ_OK;
+}
+
+static az_result append_string(az_json_writer* json_writer, void* value)
+{
+  return az_json_writer_append_string(json_writer, *(az_span*)value);
+}
+
+static az_result append_json_token(az_json_writer* json_writer, void* value)
+{
+  az_json_token value_token = *(az_json_token*)value;
+
+  double value_as_double;
+
+  switch (value_token.kind)
+  {
+    case AZ_JSON_TOKEN_NUMBER:
+      AZ_RETURN_IF_FAILED(az_json_token_get_double(&value_token, &value_as_double));
+      AZ_RETURN_IF_FAILED(
+          az_json_writer_append_double(json_writer, value_as_double, DOUBLE_DECIMAL_PLACE_DIGITS));
+      break;
+    case AZ_JSON_TOKEN_STRING:
+      AZ_RETURN_IF_FAILED(az_json_writer_append_string(json_writer, value_token.slice));
```

Comment:
As a follow up to our conversation, until [this](https://github.com/Azure/azure-sdk-for-c/issues/934) feature is in, I would first call `az_json_token_get_string(...)` before passing it to `az_json_writer_append_string`.

`az_json_writer_append_string` does escaping, so passing in a raw token slice which hasn't been unescaped, would change the output JSON.

For example: `"foo\u0061"` will become `"foo\\u0061"` when passed in to the writer. Embedded quotes will be a bigger problem as they would result in invalid JSON: `"Hi \"Dane\""` becomes `"Hi \\"Dane\\""` and now the string ends right after Hi.

On the other hand: getting the unescaped string `fooa` and then writing it out as `fooa` is OK because those are semantically equivalent strings.